components/util/src/datap/mod.rs

```diff
@@ -0,0 +1,186 @@
+// Data provider trait definitions
+
+pub mod decimal;
+
+mod helpers;
+
+use core::ops::Deref;
+use helpers::CloneableAny;
+use std::any::Any;
+use std::any::TypeId;
+use std::borrow::Borrow;
+use std::borrow::BorrowMut;
+use std::borrow::Cow;
+use std::error::Error;
+use std::fmt::{self, Debug, Display};
+use std::prelude::v1::*;
+
+/// A top-level collection of related data keys.
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum Category {
+    Undefined = 0,
+    Decimal = 1,
+    PrivateUse = 4096,
+}
+
+impl Display for Category {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
+/// A specific key within a category.
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum Key {
+    Undefined,
+    Decimal(decimal::Key),
+    PrivateUse(u32),
+}
+
+impl Display for Key {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
+/// A struct to request a certain hunk of data from a data provider.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Request {
+    // TODO: Make this a Locale instead of a String
+    pub locale: String,
+    pub category: Category,
+    pub key: Key,
+
+    // For now, the request payload is a string. If a more expressive type is needed in the
+    // future, it should implement PartialEq and Clone. Consider using a concrete type, rather
+    // than a detour through Any (like in Response), which complicates the code.
+    pub payload: Option<Cow<'static, str>>,
+}
+
+/// A response object containing a data hunk ("payload").
+#[derive(Debug, Clone)]
+pub struct Response {
+    // TODO: Make this a Locale instead of a String
+    pub data_locale: String,
```

Comment:
> Strictly speaking, public fields become part of the public API and are essentially ~forever.

Not forever. I think it's an important distinction for me and Shane due to our work on ECMA402. In ECMA402 context public API is actually forever.
Here, the public API is until the next major revision. So if we release ICU4X 0.1, we can change it in 0.2. When we release 1.0, we can change that in 2.0 and so on.

I don't say it to diminish the value of designing a clean API, just to clarify that we have wiggle room to experiment and iterate over the API.

Reply:
Also, the `best practice guide` part about public fields has been under discussion. I disagree with the absolute tone of that assumption. I believe that if the field operation is independent of any other state and doesn't affect it and doesn't require validation or canonicalization, public fields are perfectly reasonable API.

I'm not sure if this is the case here yet, so my comment is more of the assumption that public fields are a bad practice in general.

In this particular case, I assume that it's a WIP and public fields are useful for prototyping until you have the right data model and you can wrap them in getter/setter if needed.

Reply:
Thanks for explaining. (And given that your explanation seems familiar, I may not have paid enough attention at some point prior.) I am not dogmatic about this; rather this is my 
default position (based on personal priors), but it is open to corrections depending on specific project's needs.  If you say that from your experience there is a practical "out" that has worked in practice in the past for a project of this kind, I'm fine with that.

Reply:
> I am not dogmatic about this; rather this is my default position (based on personal priors)

I think it's a general good guidance, and most of the time we'll want to avoid public fields. In particular is something is set by the constructor and meant not to be touched after, it shouldn't be public, so I think your gut feeling was right in this case!

Was just reacting to the generalistic nature of your comment :)

Reply:
To sum it up, if you're setting a non-public field and then writing:
```rust
    pub fn field(&self) -> &Field { self.field };
    pub fn set_field(&mut self, f: Field) { self.field = Field };
```

then public field is a perfectly fine choice instead. If you have any operations in getter/setter, or if you don't want to provide the setter, then you shouldn't use public field.